src/TextRenderer.vala

```diff
@@ -126,12 +126,24 @@ namespace Marlin {
                 x_offset += border_radius;
             }
 
+            var provider = new Gtk.CssProvider ();
```

Comment:
I think this could be a little bit more optimized. Instead of creating a new provider every render we can only create it when the result of `contrasting_foreground_color` changed. The provider object also could be also created within the constructor and here we could only call `load_from_data`, we should also probably remove the existing provider if it exists.